File: Neetcode 150/Array s-Hashing/encode and decode string/encodeAndDecode.js
class Solution {
  // Method to encode a list of strings into a single string
  encode(strs) {
    let res = "";
    for (let val of strs) {
      res += val.length + "#" + val;
    }
    return res;
  }

  // Method to decode a single string back into a list of strings
  decode(s) {
    let res = [];
    let i = 0;

    while (i < s.length) {
      let j = i;
      while (s[j] !== "#") {
        j++;
      }
      let length = parseInt(s.substring(i, j));
      i = j + 1;
      j = i + length;
      res.push(s.substring(i, j));
      i = j;
    }

    return res;
  }
}

File: Neetcode 150/Array s-Hashing/encode and decode string/encodeAndDecode.py
class Solution:

    def encode(self, strs: List[str]) -> str:
        res = ''
        for val in strs:
            res += str(len(val)) + '#' + val
        return res


    def decode(self, s: str) -> List[str]:
        res, i = [], 0
        
        while i < len(s):
            j = i
            while s[j] != '#':
                j += 1
            length = int(s[i:j])
            i = j + 1
            j = i + length
            res.append(s[i:j])
            i = j
        
        return res
